drivers/sensor/mockup/battery/battery.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <kernel.h>
+#include <sensor.h>
+
+struct battery_data {
+	u8_t battery_lvl;
+};
+
+static struct device *device;
+static int battery_init(struct device *dev);
+static sensor_trigger_handler_t trigger_callback;
+
+static void trigger_fire(struct k_work *work)
+{
+	struct sensor_trigger trigger = {
+		.chan = SENSOR_CHAN_SOC,
+		.type = SENSOR_TRIG_DATA_READY
+	};
+
+	trigger_callback(device, &trigger);
+}
+
+K_WORK_DEFINE(timer_work, trigger_fire);
+
+static void timer_expire(struct k_timer *dummy)
+{
+	k_work_submit(&timer_work);
+}
+
+K_TIMER_DEFINE(trigger_timer, timer_expire, NULL);
+
+static int battery_init(struct device *dev)
+{
+	device = dev;
+	return 0;
+}
+
+static int battery_trigger_set(struct device *dev,
+			       const struct sensor_trigger *trig,
+			       sensor_trigger_handler_t handler)
+{
+	if (trig->type != SENSOR_TRIG_DATA_READY ||
+	    trig->chan != SENSOR_CHAN_SOC) {
+		return -EINVAL;
+	}
+
+	trigger_callback = handler;
+
+	/* start periodic timer */
+	k_timer_start(&trigger_timer,
+		      K_MSEC(CONFIG_BATTERY_SENSOR_TRIG_PERIOD),
+		      K_MSEC(CONFIG_BATTERY_SENSOR_TRIG_PERIOD));
+
+	return 0;
+}
+
+static int battery_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	if (chan != SENSOR_CHAN_ALL &&
+	    chan != SENSOR_CHAN_SOC) {
+		return -ENOTSUP;
+	}
+
+	struct battery_data *data = dev->driver_data;
+
+	data->battery_lvl = data->battery_lvl > CONFIG_BATTERY_SENSOR_LVL_MIN
+				? data->battery_lvl - 1
+				: CONFIG_BATTERY_SENSOR_LVL_MAX;
```

Comment:
This is a mockup that simulates a decrease in the battery level, when the minimum level is reached, the battery level resets. This maintains the behavior of the previous BAS implementation.

Reply:
It would be nice to see some comment like:
/* Simulating a decrease in the battery level. */

This makes it obvious. Or don't. That may be only me who feels a little lost here as I am still not totally in the architecture and code structure.

Reply:
Id change the lvl on the work function not on fetch otherwise if someone fetch it again t would change the level without signalling which imo is not correct since it would not honor CONFIG_BATTERY_SENSOR_TRIG_PERIOD.